### PR TITLE
Changes to add Account support

### PR DESCRIPTION
**Pull request by piotrm50**
# Description of change

This PR introduces a large number of changes that add support for using AccountOutput in a transaction, in place of AliasOutput. 
It also extends VM to support Transaction validation when creating, destroying or modifying an AccountOutput.

## Type of change

Choose a type of change, and delete any options that are not relevant.
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Implemented unit tests.
This code is also used in iota-core software which uses it in it's own test cases. 

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests using ginkgo that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -99,6 +101,12 @@ type ProtocolParameters struct {
 	GenesisUnixTimestamp uint32 `serix:"6,mapKey=genesisUnixTimestamp"`
 	// SlotDurationInSeconds defines the duration of each slot in seconds.
 	SlotDurationInSeconds uint8 `serix:"7,mapKey=slotDurationInSeconds"`
+	// ManaGenerationRate is the amount of potential Mana generated by 1 IOTA in 1 slot.
+	ManaGenerationRate uint8 //`serix:"8,mapKey=manaGenerationRate"`
```
Comment (alexsporn) [inline api.go:105]: why are the serix tags commented out?

File: api.go [hunk 2]
```diff
@@ -47,6 +47,8 @@ func LatestAPI(protoParams *ProtocolParameters) API {
 }
 
 // calls the internally instantiated API to encode the given object.
+//
+//nolint:unparam
```
Comment (alexsporn) [inline api.go:51]: why unparam here?

File: api_test.go [hunk 3]
```diff
@@ -86,8 +86,9 @@ func TestProtocolParametersJSONMarshalling(t *testing.T) {
 		TokenSupply:           1234567890987654321,
 		GenesisUnixTimestamp:  1681373293,
 		SlotDurationInSeconds: 10,
+		MaxCommitableAge:      10,
```
Comment (alexsporn) [inline api_test.go:89]: Add the new mana values to this test

File: bic.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package iotago
+
+type BlockIssuanceCredit struct {
+	AccountID
```
Comment (alexsporn) [inline bic.go]: Give the variable a name

File: builder/transaction_builder.go [hunk 5]
```diff
@@ -41,14 +42,17 @@ type TxInput struct {
 	// The ID of the referenced input.
 	InputID iotago.OutputID `json:"inputID"`
 	// The output which is used as an input.
-	Input iotago.Output `json:"input"`
+	Input iotago.OutputWithCreationTime `json:"input"`
```
Comment (alexsporn) [inline builder/transaction_builder.go]: I think this needs to be reverted, the `OutputWithCreationTime` is something just the VM needs

File: chain.go [hunk 6]
```diff
@@ -28,9 +26,17 @@ const (
...
+type ChainOutputWithCreationTime struct {
+	Output       ChainOutput
+	CreationTime SlotIndex
+}
+
 // ChainOutputSet is a map of ChainID to ChainOutput.
 type ChainOutputSet map[ChainID]ChainOutput
 
+// ChainInputSet is a map of ChainID to ChainOutputWithCreationTime.
+type ChainInputSet map[ChainID]*ChainOutputWithCreationTime
```
Comment (alexsporn) [inline chain.go]: maybe move these types to the VM because it is only needed there

File: feat_blockissuer.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	BlockIssuerKeys BlockIssuerKeys `serix:"0,mapKey=blockIssuerKeys,lengthPrefixType=uint8"`
+	ExpirySlot      SlotIndex       `serix:"1,mapKey=expirySlot"`
+}
+
+func (s *BlockIssuerFeature) Clone() Feature {
+	return &BlockIssuerFeature{BlockIssuerKeys: s.BlockIssuerKeys, ExpirySlot: s.ExpirySlot}
+}
+
+func (s *BlockIssuerFeature) VBytes(rentStruct *RentStructure, _ VBytesFunc) VBytes {
+	// TODO: add factor for block issuer keys (higher than regular keys factor).
```
Comment (alexsporn) [inline feat_blockissuer.go:32]: what about this TODO?

File: input.go [hunk 8]
```diff
@@ -66,48 +68,84 @@ type InputsSyntacticalValidationFunc func(index int, input Input) error
...
+			if err := f(i, input); err != nil {
 				return err
 			}
 		}
 	}
+
 	return nil
 }
+
+// TODO: does this belong here?
```
Comment (alexsporn) [inline input.go]: no, move to the VM

File: output.go [hunk 9]
```diff
@@ -207,9 +211,17 @@ func MustOutputIDFromHex(hexStr string) OutputID {
 	return outputID
 }
 
+type OutputWithCreationTime struct {
```
Comment (alexsporn) [inline output.go]: Move this to the VM

File: vm/stardust/vm_stardust.go [hunk 10]
```diff
@@ -161,71 +186,186 @@ func aliasStateSTVF(current *iotago.AliasOutput, next *iotago.AliasOutput, vmPar
...
+
+	} else if nextBIFeat != nil {
+		// if the block issuer feature has expired, it must either be removed or expiry extended.
+		if nextBIFeat.ExpirySlot != 0 && nextBIFeat.ExpirySlot < txSlotIndex+iotago.SlotIndex(vmParams.External.ProtocolParameters.MaxCommitableAge) {
+			return fmt.Errorf("%w: block issuer feature expiry set too soon", iotago.ErrInvalidBlockIssuerTransition)
+		}
+	}
+
+	// the Mana on the account on the input side must not be moved to any other outputs or accounts.
+	// TODO: include outputs with ManaLock Conditions in this check
```
Comment (alexsporn) [inline vm/stardust/vm_stardust.go]: ManaLock?

File: vm/vm.go [hunk 11]
```diff
@@ -403,6 +432,22 @@ func ExecFuncSenderUnlocked() ExecFunc {
 	}
 }
 
+// ExecFuncBalancedMana validates that Mana is balanced from the input/output side.
+// TODO: Return Mana according to StorageDepositReturnUnlockCondition(s)?
```
Comment (alexsporn) [inline vm/vm.go:436]: what about the TODO?

File: vm/stardust/vm_stardust.go (hunk 10 above)
Comment (cyberphysic4l) [reply vm/stardust/vm_stardust.go]: This is done now below. The TODO comment should be removed.

File: mana_decay_provider.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+	return uint64((slotIndex-1)/p.slotsPerDecayIndex) + 1
+}
+
+func (p *ManaDecayProvider) StoredManaWithDecay(storedMana uint64, slotIndexLast SlotIndex, slotIndexNew SlotIndex) uint64 {
+	decayIndexLast := p.decayIndex(slotIndexLast)
+	decayIndexNew := p.decayIndex(slotIndexNew)
+
+	if decayIndexLast > decayIndexNew {
+		panic(fmt.Sprintf("the last decay index was bigger than the new decay index: %d > %d", decayIndexLast, decayIndexNew))
```
Comment (alexsporn) [inline mana_decay_provider.go]: Remove this panic and instead return an error and fail validation.

File: mana_decay_provider.go [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+	return p.decay(storedMana, decayIndexNew-decayIndexLast)
+}
+
+func (p *ManaDecayProvider) PotentialManaWithDecay(deposit uint64, slotIndexLast SlotIndex, slotIndexNew SlotIndex) uint64 {
+	decayIndexLast := p.decayIndex(slotIndexLast)
+	decayIndexNew := p.decayIndex(slotIndexNew)
+
+	if decayIndexLast > decayIndexNew {
+		panic(fmt.Sprintf("the last decay index was bigger than the new decay index: %d > %d", decayIndexLast, decayIndexNew))
```
Comment (alexsporn) [inline mana_decay_provider.go]: Same as above

File: feat_blockissuer.go (hunk 7 above)
Comment (alexsporn) [reply feat_blockissuer.go:32]: Tracked in https://github.com/iotaledger/iota-core/issues/137

File: vm/vm.go (hunk 11 above)
Comment (alexsporn) [reply vm/vm.go:436]: https://github.com/iotaledger/iota-core/issues/138

**Pull request merged**